### PR TITLE
Use asyncio instead of gevent

### LINKED ISSUE #29: Communicate with Minio asynchronously
Issue produced by #26

### LINKED ISSUE #35: Rethink async architecture
I think that good things might happen if we ditch gevent and use `asyncio` from the python (3.5+) standard library. Some semi-random thoughts on that follow:

1. We could use an async s3 client (like [this](https://github.com/aio-libs/aiobotocore) or [this](https://github.com/uktrade/aioftps3)) to download payloads of queued jobs in advance while a sheep is eating something. This would be a good opportunity to abstract Minio stuff away to a service so that we can  swap file storage implementations if needed.
2. We might have to rethink how the shepherd API works. Currently, we use a WSGI server from gevent that takes care of everything for us. Off the top of my head, we could replace Flask with [Sanic](https://github.com/huge-success/sanic) - a fast, async, pure-python HTTP server with an API similar to Flask. We would have to make some adjustments to apistrap though (and it's hard to tell how hard it would be). **However**, it seems that you can also call asyncio tasks from ordinary Flask routes too. That would mean we could run the API without modifications using e.g. aiohttp-wsgi.
3. PyZMQ [supports asyncio](https://pyzmq.readthedocs.io/en/latest/eventloop.html).

### PR DESCRIPTION
**Pull request by Teyras**
closes #35 
closes #29

- [x] basic functionality
- [x] tests
- [x] async communication with Minio (aiobotocore)
- [x] CORS support (https://github.com/aio-libs/aiohttp-cors#usage)

### CODE REVIEW

**Comment by Teyras** [comment]
This requires the `aiohttp` branch of `apistrap`. The dummy model seems to work and we can fetch results. Minio is not yet acessed asynchronously and tests need some adjustments.

**Comment by blazekadam** [comment]
@Teyras you may want to resolve the conflicts rather sooner than later :)

**Comment by blazekadam** [comment]
Is this ready to be reviewed btw?

**Comment by Teyras** [comment]
See the checklist. IMO, it will be better if you review it now and then again, when I finish the remaining tasks.

**Comment by Teyras** [comment]
@blazekadam feel free to review this. When I finish rehauling job status files, I'll either make a PR into this branch or into dev.

File: requirements.txt [hunk 1]
```diff
@@ -15,5 +15,6 @@ git+https://github.com/iterait/apistrap.git
 # testing
 pytest
 pytest-mock
```
Comment (blazekadam) [inline requirements.txt]: Flask_CORS and Flask are still required?|

File: shepherd/api/app.py [hunk 2]
```diff
@@ -29,8 +29,8 @@ def error_handler(http_code, error: AppError):
     return (ErrorResponse({"message": str(error)})), http_code
 
 
-def create_app():
-    app = web.Application(debug=os.getenv('DEBUG', False))
+def create_app(debug=None):
```
Comment (blazekadam) [inline shepherd/api/app.py]: docs & types?

File: shepherd/api/app.py [hunk 3]
```diff
@@ -29,8 +29,8 @@ def error_handler(http_code, error: AppError):
...
+    app = web.Application(debug=debug if debug is not None else os.getenv('DEBUG', False))
```
Comment (blazekadam) [inline shepherd/api/app.py]: `debug or False`?

File: tests/api/test_get_result.py [hunk 4]
```diff
@@ -24,29 +24,38 @@ def job_failed(minio: Minio, bucket):
...
 
+    response = await client.get("/jobs/{}/result/payload.json".format(job_id))
+    assert response.status == 202
 
-def test_get_result_error(job_failed, client):
+
+async def test_get_result_error(job_failed, aiohttp_client, app):
     job_id = job_failed
-    response = client.get("/jobs/{}/result/payload.json".format(job_id))
+    client = await aiohttp_client(app)
```
Comment (blazekadam) [inline tests/api/test_get_result.py]: those awaits are indeed necessary?

File: tests/shepherd/test_shepherd.py [hunk 5]
```diff
@@ -21,13 +21,11 @@ def test_shepherd_init(valid_config: ShepherdConfig, minio):
     # test sheep getter
     with pytest.raises(UnknownSheepError):
         _ = shepherd['UnknownSheep']
-    shepherd.close()
```
Comment (blazekadam) [inline tests/shepherd/test_shepherd.py]: 👍

File: setup.py [hunk 6]
```diff
@@ -28,7 +28,7 @@
       setup_requires=['pytest-runner'],
       tests_require=['pytest', 'pytest-mock', 'pytest-aiohttp'],
       install_requires=['click', 'aiohttp', 'simplejson', 'pyzmq', 'gevent', 'PyYaml', 'requests', 'minio',
-                        'schematics', 'emloop', 'apistrap'],
+                        'schematics', 'emloop', 'apistrap', 'aiohttp-cors'],
```
Comment (blazekadam) [inline setup.py]: no need to somehow set up the CORS in the actual code?

File: tests/utils/test_storage_utils.py [hunk 7]
```diff
@@ -77,13 +77,13 @@ def test_minio_connectivity(minio: Minio):
     assert not minio_object_exists(minio, bucket, 'another/file.txt')
 
 
-async def test_minio_push_empty(storage: MinioStorage, bucket, job_dir, caplog):
+async def test_minio_push_empty(storage: MinioStorage, bucket, job_dir, caplog, minio):
```
Comment (blazekadam) [inline tests/utils/test_storage_utils.py]: why it is needed to use the `minio` fixture explicitly?

**Comment by blazekadam** [comment]
Tests are failing (and getting stuck) locally with several errors `'Application' has no attribute 'response_class'` can you advise?

**Comment by Teyras** [comment]
@blazekadam off the top of my head, old dependencies? I suspect apistrap.

File: tests/utils/test_storage_utils.py (hunk 7 above)
Comment (Teyras) [reply tests/utils/test_storage_utils.py]: possibly a result of some experimentation when CI was failing... will investigate.

File: tests/api/test_get_result.py (hunk 4 above)
Comment (Teyras) [reply tests/api/test_get_result.py]: yeah, they're never voluntary :) you can either await them or explicitly fire the coroutine into the background...

File: shepherd/api/app.py (hunk 3 above)
Comment (Teyras) [reply shepherd/api/app.py]: That would bypass the env variable

File: requirements.txt (hunk 1 above)
Comment (Teyras) [reply requirements.txt]: touché

File: shepherd/api/app.py (hunk 2 above)
Comment (Teyras) [reply shepherd/api/app.py]: if you insist...

File: shepherd/runner/runner_entry_point.py [hunk 8]
```diff
@@ -51,8 +52,12 @@ def run() -> None:
...
+    await runner.process_all()
+
+
+def run():
+    asyncio.run(main())  # pragma: no cover
 
 
 if __name__ == '__main__':
-    run()  # pragma: no cover
+    run()
```
[the remark this answers is not in the document]
Comment (Teyras) [reply shepherd/runner/runner_entry_point.py]: Addressed in later commits, I believe.

File: shepherd/sheep/base_sheep.py [hunk 9]
```diff
@@ -1,10 +1,10 @@
 import abc
 import logging
 from typing import List, Optional, Dict
+from asyncio import Queue
 
-import zmq.green as zmq
+import zmq.asyncio
```
[the remark this answers is not in the document]
Comment (Teyras) [reply shepherd/sheep/base_sheep.py]: Will investigate, there's weird stuff going on in the zmq package.

File: shepherd/api/views.py [hunk 10]
```diff
@@ -26,14 +28,14 @@ def check_job_exists(minio, job_id):
         raise StorageError('Failed to check minio bucket `{}`'.format(job_id)) from me
 
 
-def create_shepherd_blueprint(shepherd: Shepherd, minio: Minio):
-    api = Blueprint('shepherd', __name__)
+def create_shepherd_routes(shepherd: Shepherd, minio: Minio):
+    api = web.RouteTableDef()
 
-    @api.route('/start-job', methods=['POST'])
+    @api.post('/start-job')
```
[the remark this answers is not in the document]
Comment (Teyras) [reply shepherd/api/views.py]: Courtesy of `aiohttp` :)

File: shepherd/api/app.py [hunk 11]
```diff
@@ -31,20 +31,18 @@ def error_handler(http_code, error: AppError):
...
-    return jsonify(response.to_primitive()), http_code
+    return web.Response(text=json.dumps(response.to_primitive()), content_type='application/json'), http_code
 
 
-def create_app(name: str):
-    app = Flask(name)
-    CORS(app, expose_headers=["Content-Disposition"], send_wildcard=True, origins=[])
+def create_app():
+    app = web.Application(debug=os.getenv('DEBUG', False))
+    # TODO CORS(app, expose_headers=["Content-Disposition"], send_wildcard=True, origins=[])
```
[the remark this answers is not in the document]
Comment (Teyras) [reply shepherd/api/app.py]: This is sorted out in later commits.

**Pull request merged**